pipeline/acquire_decam_refs.py

```diff
@@ -0,0 +1,554 @@
+import numpy as np
+
+from astropy.io import fits
+
+from models.base import SmartSession
+from models.instrument import get_instrument_instance
+from models.provenance import Provenance
+from models.decam import DECam
+from models.image import Image
+from models.exposure import Exposure
+
+from pipeline.data_store import DataStore
+from pipeline.top_level import Pipeline
+from pipeline.preprocessing import Preprocessor
+from pipeline.ref_maker import RefMaker
+
+import util.radec
+from util.logger import SCLogger
+
+
+class DECamRefFetcher:
+    def __init__( self, ra, dec, filter, chips=None, numprocs=10,
+                  min_exptime=None, max_seeing=1.2, min_depth=None,
+                  min_mjd=None, max_mjd=None, min_per_chip=9,
+                  only_local=False, no_fallback=False, full_report=False ):
+        self.numprocs = numprocs
+        self.ra = ra
+        self.dec = dec
+        self.filter = filter
+        self.min_exptime = min_exptime
+        self.max_seeing = max_seeing
+        self.min_depth = min_depth
+        self.min_mjd = min_mjd
+        self.max_mjd = max_mjd
+        self.min_per_chip = min_per_chip
+        self.only_local = only_local
+        self.no_fallback = no_fallback
+        self.full_report = full_report
+
+        # NOTE -- empirically, the NOIRLab pipeline seems to be more optimistic
+        #   about both seeing and depth that what our pipeline comes up with.
+        #   when searcing noirlab, reduce the seeing cut by seeingfac,
+        #   and add limmagoff to the limiting magnitude
+        self.seeingfac = 1.06
+        self.limmagoff = 0.5
+
+        self.decam = get_instrument_instance( 'DECam' )
+        knownchips = self.decam.get_section_ids()
+        if chips is not None:
+            if not isinstance( chips, list ):
+                raise TypeError( f"DECamRefFetcher: chips must be a list, not a {type(chips)}" )
+            if not set(chips).issubset( set(knownchips ) ):
+                raise ValueError( f"DECamRefFetcher: passed chips {chips}, which includes unknown section ids" )
+            self.chips = chips
+        else:
+            self.chips = knownchips
+
+        # Need a RefMaker for probing the database.
+        # For that, need a temporary preprocessor to figure out the provenance
+        preprocessor = Preprocessor( preprocessing='noirlab_instcal',
+                                     calibset='externally_supplied',
+                                     flattype='externally_supplied',
+                                     steps_required=['overscan', 'linearity', 'flat', 'fringe'] )
+        preprocprov = Provenance( process='preprocessing',
+                                  parameters=preprocessor.pars.get_critical_pars(),
+                                  upstreams=[ self.decam.get_exposure_provenance( proc_type='instcal' ) ],
+                                 )
+        preprocprov.insert_if_needed()
+
+        # TODO : get some of these values from config
```

Comment:
No, not in this PR.  This whole file is in a bit of an odd situation, since I really wrote it just for one survey (the DECAT WFD survey Peter ran this last semester), but realized it might be useful elsewhere.  Thus, I didn't write tests for it, but then put it in pipeline instead of hacks/rknop.